### PR TITLE
Data ext seq mask, use non-zero sizes for padded ext dyn size

### PR DESCRIPTION
**Pull request by albertz**
size_ext might have invalid (zero) sizes when it itself has some padding, e.g. when its own shape is dynamic.

A zero size can lead to problems in some cases, e.g. in SoftmaxOverSpatialLayer, when everything is masked to -inf, it results in nan, and this likely produces nan in backprop or elsewhere.

Thus, mask size_ext itself, and set the padded values to 1.

---

This is somewhat an unrelated or orthogonal follow-up to #1019, resulting from [this config](https://gist.github.com/albertz/3de318fca54b76518e24cde024913585).

@robin-p-schmitt This probably should fix the problem you showed me with nan when you disable attention dropout.

### CODE REVIEW

**Comment by albertz** [comment]
> I wonder if `1` is really the best pad value here. I think I would rather use the full unpadded seq length as mask value. `1` seems a bit arbitrary and may produce very unexpected results. E.g. if the first time frame is `-inf` in the data, and then the seq length is masked to `1`, this will also produce a `NaN` gradient after a softmax. So using the entire unpadded seq length in that case gives the highest chance to have at least one finite input frame.

Right, using `max_idx` is a better idea.

> How specific is this to `SoftmaxOverSpatialLayer`,

This is the case where I stumbled upon this problem.

This is classic attention (but over a dyn_size_ext of shape [B,S]), i.e. softmax over spatial, then dot layer. The softmax would produce nan for any seq size = 0 because softmax(-inf) = nan.

We had the strange case that we also had dropout on the attention weights, and with dropout, all was fine, but once we removed the dropout, we got nan.

I figured out that dropout resulted in removing the padding info (`set_padding_info`), so then our new `DotLayer` now again did masking, just to be sure, and replaced all the nans by 0, and then all was fine. But when there is no dropout, it would not do the masking again, and the nan will result in further problems in backprop or elsewhere.

> or are other layers maybe also a problem?

I could imagine but I didn't really check all the layers.

> We have #576 about NaN gradients after `SoftmaxOverSpatialLayer` in general.

Ah, I forgot about this issue. Yea, this is related to this PR.

> I think a nicer fix would be to change `SoftmaxOverSpatialLayer` to always mask away sequences for queries where all keys are invalid with something finite before applying the softmax. I.e. to fix #576 directly.

I don't remember the details of the issue anymore. Is this about invalid data in the padded area?

So this PR here fixes invalid seq lens in the padded area. So this is one part (or already the whole thing) to fix it, right? Is there anything missing then?

But anyway, I think this PR is also useful as-is, or not?

**Comment by albertz** [comment]
@Zettelkasten I changed the implementation a bit. `mask_dyn_seq_len_nd` got extended a bit. Can you check if this looks ok?

File: returnn/tf/util/basic.py [hunk 1]
```diff
@@ -250,26 +250,44 @@ def set_padding_info(x, dim, pad_value):
...
   :return: masked x
   :rtype: tf.Tensor
   """
+  # Filter out some axes which should not be used for masking.
+  axes_ = []
+  for axis in axes:
+    tag = x.dim_tags[axis]
+    assert tag.dyn_size_ext
+    # It only makes sense to apply for this axis if the dyn size dims are all existing in x itself.
+    # E.g. if the dyn_size_ext shape is [B] but the shape of x is just [T] (without B),
```
Comment (Zettelkasten) [inline returnn/tf/util/basic.py:264]: Wait, isn't this very dangerous / wrong?
In this case, you'd want the output of the layer to have this implicit dims (`B` in your example) as concrete dims in the output. As in https://github.com/rwth-i6/returnn/issues/721. Right?
We e.g. had a bug with this before in https://github.com/rwth-i6/returnn/issues/717.

**Comment by Zettelkasten** [comment]
> > We have #576 about NaN gradients after `SoftmaxOverSpatialLayer` in general.
> 
> Ah, I forgot about this issue. Yea, this is related to this PR.
> 
> > I think a nicer fix would be to change `SoftmaxOverSpatialLayer` to always mask away sequences for queries where all keys are invalid with something finite before applying the softmax. I.e. to fix #576 directly.
> 
> I don't remember the details of the issue anymore. Is this about invalid data in the padded area?

Yes it is. It is actually kind of the same case as here, where because of masking, all keys for one specific query would be masked away with -inf, resulting in NaN after the softmax.
Here, this masking is because of the seq len mask. In #576, more generally, the masking was done by me explicitly using other layers because I wanted a more complex mask: E.g. I masked away all keys with key=query position. But I tried quite a few of these masks.
Just like in this issue, I didn't really think about what my mask did for invalid query positions, and if I was unlucky, all keys would be masked away to -inf for one key, resulting in the NaN softmax output + gradients.

> So this PR here fixes invalid seq lens in the padded area. So this is one part (or already the whole thing) to fix it, right? Is there anything missing then?

It's fixed for the case where this masking is performed implicitly via the seq len mask, but not when the input to the softmax layer already has -inf for all keys for some query.

> But anyway, I think this PR is also useful as-is, or not?

Yes, certainly.

**Comment by albertz** [comment]
> > So this PR here fixes invalid seq lens in the padded area. So this is one part (or already the whole thing) to fix it, right? Is there anything missing then?
> 
> It's fixed for the case where this masking is performed implicitly via the seq len mask, but not when the input to the softmax layer already has -inf for all keys for some query.

I see.

But I would keep that for a separate PR / change. Also, masking the dyn seq size is cheap and easy to do (it should not really cost anything). Handling the other issue will be more involved and I cannot really imagine any solution which would not at least add some overhead.

File: returnn/tf/util/basic.py (hunk 1 above)
Comment (albertz) [reply returnn/tf/util/basic.py:264]: The usage of `mask_dyn_seq_len_nd` is only in `DotLayer` and in `Data.copy_masked` which itself is only used for the ext dyn size in `Data.get_sequence_mask_broadcast`. 

In `DotLayer`, we already check for that.

In `Data.get_sequence_mask_broadcast` for the ext dyn size, it should not matter. We just want to make sure that there are no invalid sizes (zero sizes).

You are right, it could be wrong in other cases. Currently there are no other cases and I think the current cases are both correct.

So, what should we do? Maybe just add a big warning comment with this info there?
